src/ray/core_worker/transport/actor_task_submitter.cc

```diff
@@ -227,30 +227,38 @@ Status ActorTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
         },
         "ActorTaskSubmitter::SubmitTask");
   } else {
-    // Do not hold the lock while calling into task_finisher_.
-    task_finisher_.MarkTaskCanceled(task_id);
-    rpc::ErrorType error_type;
-    rpc::RayErrorInfo error_info;
-    {
-      absl::MutexLock lock(&mu_);
-      const auto queue_it = client_queues_.find(task_spec.ActorId());
-      const auto &death_cause = queue_it->second.death_cause;
-      error_info = GetErrorInfoFromActorDeathCause(death_cause);
-      error_type = error_info.error_type();
-    }
-    auto status = Status::IOError("cancelling task of dead actor");
-    // No need to increment the number of completed tasks since the actor is
-    // dead.
-    bool fail_immediately =
-        error_info.has_actor_died_error() &&
-        error_info.actor_died_error().has_oom_context() &&
-        error_info.actor_died_error().oom_context().fail_immediately();
-    GetTaskFinisherWithoutMu().FailOrRetryPendingTask(task_id,
-                                                      error_type,
-                                                      &status,
-                                                      &error_info,
-                                                      /*mark_task_object_failed*/ true,
-                                                      fail_immediately);
+    // Post to the event loop to maintain the async nature of
+    // SubmitTask and avoid issues like
+    // https://github.com/ray-project/ray/issues/47606.
+    io_service_.post(
```

Comment:
Ehh, we do have code that relies on the behavior that this piece of code is executed immediately instead of being posted to the event loop:

```
def is_shutdown(self) -> bool:
        """Return whether the proxy actor is shutdown.

        If the actor is dead, the health check will return RayActorError.
        """
        try:
            ray.get(self._actor_handle.check_health.remote(), timeout=0)
        except RayActorError:
            # The actor is dead, so it's ready for shutdown.
            return True

        # The actor is still alive, so it's not ready for shutdown.
        return False
```